sway-lsp/src/core/session.rs

```diff
@@ -109,8 +122,15 @@ impl Session {
     }
 
     pub fn parse_project(&self, uri: &Url) -> Result<Diagnostics, LanguageServerError> {
-        self.token_map.clear();
-        self.runnables.clear();
+        // Acquire a permit to parse the project. If there are none available, exit early.
```

Comment:
I think we might want to approach this differently. If we return here, then the workspace is out of sync with the server. Ideally, we would cancel any on going work and reparse with the most up to date information from the client. 

This definitely isn't a trivial thing to implement. Ideally, we would be able to take advantage of the [cancelRequest](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#cancelRequest) part of the LSP spec to drop any current work. Unfortunately, `parse_project` is getting triggered by `didOpen` which is a notification and not an LSP request. It looks like `rust-analyzer` also bumped into this problem with notifications, see [666 ](https://github.com/microsoft/language-server-protocol/issues/666) and [584](https://github.com/microsoft/language-server-protocol/issues/584)

There is also a really good discussion [here](https://github.com/ebkalderon/tower-lsp/issues/284) where `tower-lsp` is considering ditching concurrent execution and also notes the same issues above. 

This is work that I was planning on trying to solve in #3525.